core/src/mender-utils.c

```diff
@@ -509,3 +509,161 @@ mender_utils_append_list(mender_key_value_list_t **list1, mender_key_value_list_
     *list2 = NULL;
     return MENDER_OK;
 }
+
+mender_err_t
+mender_utils_combine_unique(mender_key_value_list_t **list1, mender_key_value_list_t **list2) {
+
+    /* Append a new linked list only if the key is not already present in the first list */
+    /* This will probably mean that I need to re-iterate the first list to check for the key */
+
+    for (mender_key_value_list_t *item = *list2; item; item = item->next) {
+        bool key_exists = false;
+        for (mender_key_value_list_t *item1 = *list1; item1; item1 = item1->next) {
+            if (0 == strcmp(item->key, item1->key)) {
+                key_exists = true;
+                break;
+            }
+        }
+        if (!key_exists) {
+            if (MENDER_OK != mender_utils_create_key_value_node(item->key, item->value, list1)) {
+                mender_log_error("Unable to create key-value node");
+                return MENDER_FAIL;
+            }
+        }
+    }
+
+    *list2 = NULL;
+    return MENDER_OK;
+}
+
+mender_err_t
+mender_utils_key_value_list_delete_node(mender_key_value_list_t **list, char *key) {
+
+    mender_key_value_list_t *to_free = NULL;
+    mender_key_value_list_t *prev    = NULL;
+    mender_key_value_list_t *item    = *list;
+    while (NULL != item) {
+        if (0 == strcmp(item->key, key)) {
+            to_free = item;
+            if (NULL == prev) {
+                *list = item->next;
+                break;
+            } else {
+                prev->next = item->next;
+            }
+            break;
+        }
+        prev = item;
+        item = item->next;
+    }
+
+    free(to_free->key);
+    free(to_free->value);
+    free(to_free);
+    return MENDER_OK;
+}
+
+bool
+mender_utils_compare_wildcard(char *string, char *wildcard_string) {
```

Comment:
What do we need wildcard comparison for? `clears_provides` uses regular expressions, IIRC. Is it for something else?

Reply:
It's based on this I wrote in mender: https://github.com/mendersoftware/mender/blob/eae7540db3ba5c7fea256f217dab956a5e06bb30/src/mender-update/context/context.cpp#L236 - but this was written to avoid c++ exceptions which is not a problem here.

It only matches [wildcards](https://github.com/mendersoftware/mender-artifact/blob/master/Documentation/artifact-format-v3.md#clears_artifact_provides). I saw a thread where Joel asked about regex for zephyr, and he solved it using [picolibc](https://github.com/zephyrproject-rtos/zephyr/issues/53407), should we do something like that? I think we only  need to match wildcards.

Reply:
Oh! I see! Looks like `*` is the only thing we need to care about. [1] I didn't know that. Examples on the Internet (incl. our own docs) suggest it's a regex, but that's a different story. Let's not worry about that now. :+1: 

[1] https://github.com/mendersoftware/mender-artifact/blob/master/Documentation/artifact-format-v3.md#clears_artifact_provides